### PR TITLE
Python 3 port

### LINKED ISSUE #106: Python 3 support
Is there any interest in Python 3 support? If I submitted patches that added Python 3 support (while still keeping 2.x) would that be something that could plausibly get merged?

### PR DESCRIPTION
**Pull request by palfrey**
Fixes #106

### CODE REVIEW

**Comment by mithrandi** [comment]
I've cherry-picked the z.i advisor changes into #121; next up is probably the print statements and assert method changes. Note that we'll want at least one release with both Python 2 and Python 3 compatibility to ease the porting burden on users, so Python-2-incompatible changes need to be avoided initially.

**Comment by glyph** [comment]
2.7 is working again, and the number of errors on 3.7 is starting to look pretty tractable.  I think fixing the need to hash `_StoreIDComparer` in rollbacks will clear a lot of the remaining noise.  (We may need to reach up into `epsilon` to make Time sortable on py3 as well.)

**Comment by glyph** [comment]
filed https://github.com/twisted/epsilon/issues/38

**Comment by glyph** [comment]
@palfrey if you want to review this I can open my own PR on the same commit and you can approve / request changes there, since obviously I wrote a lot of code here

**Comment by glyph** [comment]
and hey, what the heck, it's a party, inviting @exarkun to a review as well since this'll unblock type annotations for Imaginary which might be interesting

File: axiom/item.py
Comment (glyph) [inline axiom/item.py:124]: https://bugs.python.org/issue44605

File: axiom/slotmachine.py
Comment (palfrey) [inline axiom/slotmachine.py:4]: Double-import!

File: setup.cfg
Comment (palfrey) [inline setup.cfg:15]: Should add classifiers for Python 3.x

**Review by palfrey — COMMENTED** [review]
Thanks for finishing this off!

I've noticed a few couple of things (see below), but TBH my level of knowledge of this codebase isn't enough to properly review (and hence why I got stuck in my conversion efforts). One item of slight concern is the quantity of what were explicitly unicode strings before that are now "whatever the current version of Python defaults to" strings, which is fine for 3.x but maybe less so for 2.x. Not sure how much of an effect that is?

There's a few whitespace changes as well, which although I agree with them, it might be worth running over the codebase with `black` or similar after this gets merged just to try and avoid having them in random PRs like this in the future.

**Comment by exarkun** [comment]
Thanks for your efforts here.  I don't suppose there's any chance this could be broken into more manageably sized pieces?  I'm not interested enough to make it through 2000+ lines of diff but I might look at a handful of ~100 line diffs one at a time.

**Review by glyph — APPROVED** [review]
I suspect that if anyone is still using Axiom, given the support situation for Python 2, they're probably using this branch by now. Large as it is, I'm inclined to just merge it, and move forward from there.

I've had a cursory look over the code and it looks fine. I suspect when I land it CI will break because all our supported versions are too old.

Given the relative dearth of other development going on in Axiom-land, if we need to roll this back we can always just do that, but it seems like the options at this point are "totally moribund and non-functional" or "merge an overly large branch at once".

**Pull request merged**
